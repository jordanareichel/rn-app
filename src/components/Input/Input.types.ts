import {TextInputProps} from 'react-native';

export type MaskTypes = 'cpf' | 'cnpj' | 'phone' | 'cep' | 'data' | 'currency';

export type InputProps = {
  value?: number | string;
  mask?: MaskTypes;
  label?: string;
  placeholder?: string;
  color?: string;
  maxLength?: number;
  square?: boolean;
  radius?: number;
  variant?: 'standard' | 'outlined' | 'ghost';
  maskType?: 'render' | 'value';
  disabled?: boolean;
  width?: number;
  onChangeText?: (text: any) => void;
  onKeyPress?: () => void;
  onPress?(): void;
} & TextInputProps;
