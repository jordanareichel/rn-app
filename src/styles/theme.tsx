export const theme = {
  font: {
    sm: '12',
    md: '14',
    lg: '18',
  },
  colors: {
    dark: '#000000',
    light: '#ffffff',
  },
};
