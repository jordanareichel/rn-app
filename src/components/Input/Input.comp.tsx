import React, {useCallback, useEffect, useRef} from 'react';
import {Label, TextInput, Wrapper} from './Input.styles';
import {InputProps} from './Input.types';
import {TextInput as TextInputRef} from 'react-native';

export const Input: React.FC<InputProps> = props => {
  const {
    label,
    mask,
    width,
    variant,
    color,
    value,
    radius,
    disabled,
    placeholder,
    onChangeText: onChangeTextProps,
    ...rest
  } = props;

  const refInput = useRef<TextInputRef>(null);

  useEffect(() => {
    if (refInput.current) {
      refInput.current.setNativeProps({disabled});
    }
  }, [refInput, disabled]);

  const onChangeText = useCallback(
    (text: string) => {
      switch (mask) {
        case 'cpf':
          text = text.replace(/\D/g, '');
          text = text.replace(/(\d{3})(\d)/, '$1.$2');
          text = text.replace(/(\d{3})(\d)/, '$1.$2');
          text = text.replace(/(\d{3})(\d{1,2})$/, '$1-$2');
          break;
        case 'cnpj':
          text = text.replace(/\D/g, '');
          text = text.replace(/(\d{3})(\d)/, '$1.$2');
          text = text.replace(/(\d{3})(\d)/, '$1.$2');
          text = text.replace(/(\d{3})(\d{1,2})$/, '$1-$2');
          break;
        case 'phone':
          text = text.replace(/\D/g, '');
          text = text.replace(/^(\d\d)/g, '($1)');
          text = text.replace(/^(\d\d)(\d)/g, '($1) $2');
          text = text.replace(/(\d{5})(\d)/, '$1-$2');
          break;
        case 'cep':
          text = text.replace(/\D/g, '');
          text = text.replace(/(\d{5})(\d{1,3})$/, '$1-$2');
          break;
        case 'data':
          text = text.replace(/\D/g, '');
          text = text.replace(/^(\d{2})(\d)/, '$1/$2');
          text = text.replace(/.(\d{2})(\d)/, '$1/$2');
          text = text.replace(/(\d{2})(\d)/, '$1/$2');
          break;
        case 'currency':
          if (text) {
            text = text.toString().replace(/\D/g, '');
            text =
              'R$ ' +
              text
                .toString()
                .replace(/(\d)(\d{8})$/, '$1.$2')
                .replace(/(\d)(\d{8})$/, '$1.$2')
                .replace(/(\d)(\d{2})$/, '$1,$2');
          }
          break;
        default:
          break;
      }
      onChangeTextProps && onChangeTextProps(text);
    },
    [mask, onChangeTextProps],
  );

  return (
    <Wrapper {...{width, radius, disabled, variant}}>
      <TextInput
        {...Object.assign(
          {
            value,
            onChangeText,
          },
          rest,
        )}
        ref={refInput}
        editable={!disabled}
        placeholder={placeholder}
      />
      {!!label && (
        <Label
          {...Object.assign({
            disabled,
            color,
          })}>
          {label}
        </Label>
      )}
    </Wrapper>
  );
};
