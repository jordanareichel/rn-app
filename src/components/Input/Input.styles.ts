import {Text} from '@components/Text';
import styled, {css} from 'styled-components/native';
import {InputProps} from './Input.types';

export const Wrapper = styled.View<InputProps>`
  ${({width, disabled, radius = 4, variant = ''}) =>
    css`
      border-radius: ${radius}px;
      width: ${width ? `${width}px` : '100%'};
      overflow: ${['ghost'].includes(variant) ? 'visible' : 'hidden'};
      background-color: ${!disabled ? '#DEDCDB' : 'black'};
    `}
  min-height: 56px;
  position: relative;
  border-style: solid;
  flex-direction: row;
  align-items: center;
`;

export const TextInput = styled.TextInput`
  font-size: 16px;
  flex: 1;
  top: 10px;
  left: 5px;
  z-index: 2;
  width: 100%;
  height: 100%;
  border: none;
`;

export const Label = styled(Text)`
  left: 8px;
  position: absolute;
  align-self: center;
`;
