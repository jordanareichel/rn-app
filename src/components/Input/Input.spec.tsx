import {theme} from '@styles/theme';
import React from 'react';

import TestRenderer from 'react-test-renderer';
import {ThemeProvider} from 'styled-components/native';
import {Input} from './Input.comp';

describe('Component Input', () => {
  test('renderer currently', () => {
    const tree = TestRenderer.create(
      <ThemeProvider theme={theme}>
        <Input />
      </ThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
