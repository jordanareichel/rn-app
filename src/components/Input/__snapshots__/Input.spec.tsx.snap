// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Input renderer currently 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#DEDCDB",
        "borderRadius": 4,
        "borderStyle": "solid",
        "flexDirection": "row",
        "minHeight": 56,
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      },
    ]
  }
>
  <TextInput
    editable={true}
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderColor": "black",
          "borderStyle": "solid",
          "borderWidth": 0,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontSize": 16,
          "height": "100%",
          "left": 5,
          "top": 10,
          "width": "100%",
          "zIndex": 2,
        },
      ]
    }
  />
</View>
`;
